### DISCUSSION TITLE
Instantiating a problem with multiple client time windows?

### DISCUSSION THREAD

**Discussion by ljwolf** [PyVRP/PyVRP#673] [Q&A] [answered]
Hi!

I've really enjoyed digging into `pyvrp`. It's really well designed! I've been looking for something for VRPs that is simpler than `or-tools` to add to my [introduction to spatial optimization materials](https://github.com/ljwolf/intro-to-spatialopt) when I teach. I'm building some tooling to take the output of the `pyvrp` solvers and generate a `folium` webmap and a route/stop `GeoDataFrame` complete with pandas.datetime ETAs. 

I'm currently working through an example problem with time windows where clients may close in the middle of the day. The way I'm implementing this is by building a mutually-exclusive `ClientGroup` for the multiple different periods:

```python
for i,name in enumerate(client_names):
    time_window = time_windows[i]
    if len(time_window) == 4:
        grp = model.add_client_group()
    else:
        grp = None
    c1 = model.add_client(
        x=xs[i], 
        y=ys[i], 
        delivery=demands[i], 
        tw_early=time_window[0], 
        tw_late=time_window[1], 
        name=name, 
        required=False, 
        service_duration=service_durations[i], 
        group=grp
    )
    if grp is not None:
        c2 = model.add_client(
            x=xs[i], 
            y=ys[i], 
            delivery=demands[i], 
            tw_early=time_window[2], 
            tw_late=time_window[3], 
            name=name, 
            required=False, 
            service_duration=service_durations[i], 
            group=grp
        )
```
When I do this, I sometimes get deliveries that occur between `c1.tw_late` and `c2.tw_early`. 

I notice there's no example for this case, so I wanted to confirm that 
- I've specified the client groups correctly?
- this means that the truck is assumed to wait at the target until `c2.tw_early`, do its delivery in `service_durations[i]` time, and then proceed on its route. 

Is that correct?

**Comment by N-Wouda** [comment, marked as answer]
Hi @ljwolf,

Thanks for your interest - cool to see PyVRP making its way into your teaching materials!

> * I've specified the client groups correctly?

The snippet you shared looks correct to me (I'm assuming `time_windows[i]` is either `[<tw early1>, <tw late1>]` or `[<tw early1>, <tw late1>, <tw early2>, <tw late2>]`). 

> When I do this, I sometimes get deliveries that occur between `c1.tw_late` and `c2.tw_early`.

Are the solutions being returned reported feasible by PyVRP? You can check with `res.best.is_feasible()`.

I'm assuming they are. Then there should indeed be wait time on the route, since delivery at `c2` can only happen from `c2.tw_early` onwards, so the truck has to wait when arriving there early. The route `c2` is on should have a non-zero wait time in that case, which you can check by obtaining the correct route (one of `res.best.routes()`) and calling `route.wait_duration()`.

> * this means that the truck is assumed to wait at the target until `c2.tw_early`, do its delivery in `service_durations[i]` time, and then proceed on its route.

Indeed.

**Reply by N-Wouda** [reply]
You're solving a VRP with multiple time windows. We haven't had much chance yet to determine how well our algorithm works for such VRPs. If you happen to have a comparison of PyVRP vs. for example OR-Tools, would you be willing to share the results?

**Reply by ljwolf** [reply]
Great! Yes, solved routes are feasible, hence why I'd assumed the wait time. I'll make sure to check the wait time and adjust the calculated time deltas.

Happy to share the example data once it's cleaned up!

**Reply by N-Wouda** [reply]
Great, and good luck!

**Reply by N-Wouda** [reply]
@ljwolf FYI - in the latest release we've added a `route.schedule()` member function that returns ETA, service and wait time duration for each stop.
